esphome/components/kamstrup_kmp/sensor.py

```diff
@@ -0,0 +1,158 @@
+            cv.Optional(CONF_TEMP1): sensor.sensor_schema(
+                accuracy_decimals=1,
+                device_class=DEVICE_CLASS_TEMPERATURE,
+                state_class=STATE_CLASS_MEASUREMENT,
+                unit_of_measurement=UNIT_CELSIUS,
+            ),
+            cv.Optional(CONF_TEMP2): sensor.sensor_schema(
+                accuracy_decimals=1,
+                device_class=DEVICE_CLASS_TEMPERATURE,
+                state_class=STATE_CLASS_MEASUREMENT,
+                unit_of_measurement=UNIT_CELSIUS,
+            ),
+            cv.Optional(CONF_TEMP_DIFF): sensor.sensor_schema(
+                accuracy_decimals=1,
+                device_class=DEVICE_CLASS_TEMPERATURE,
+                state_class=STATE_CLASS_MEASUREMENT,
+                unit_of_measurement=UNIT_KELVIN,
+            ),
+            cv.Optional(CONF_FLOW): sensor.sensor_schema(
+                accuracy_decimals=1,
+                device_class=DEVICE_CLASS_VOLUME,
+                state_class=STATE_CLASS_MEASUREMENT,
+                unit_of_measurement=UNIT_LITRE_PER_HOUR,
+            ),
+            cv.Optional(CONF_VOLUME): sensor.sensor_schema(
+                accuracy_decimals=1,
+                device_class=DEVICE_CLASS_VOLUME,
+                state_class=STATE_CLASS_MEASUREMENT,
+                unit_of_measurement=UNIT_CUBIC_METER,
+            ),
+            cv.Optional(CONF_CUSTOM_1): sensor.sensor_schema(
+                accuracy_decimals=1,
+                device_class=DEVICE_CLASS_EMPTY,
+                state_class=STATE_CLASS_MEASUREMENT,
+                unit_of_measurement=UNIT_EMPTY,
+            ).extend({cv.Required(CONF_COMMAND): cv.hex_uint16_t}),
+            cv.Optional(CONF_CUSTOM_2): sensor.sensor_schema(
+                accuracy_decimals=1,
+                device_class=DEVICE_CLASS_EMPTY,
+                state_class=STATE_CLASS_MEASUREMENT,
+                unit_of_measurement=UNIT_EMPTY,
+            ).extend({cv.Required(CONF_COMMAND): cv.hex_uint16_t}),
+            cv.Optional(CONF_CUSTOM_3): sensor.sensor_schema(
+                accuracy_decimals=1,
+                device_class=DEVICE_CLASS_EMPTY,
+                state_class=STATE_CLASS_MEASUREMENT,
+                unit_of_measurement=UNIT_EMPTY,
+            ).extend({cv.Required(CONF_COMMAND): cv.hex_uint16_t}),
+            cv.Optional(CONF_CUSTOM_4): sensor.sensor_schema(
+                accuracy_decimals=1,
+                device_class=DEVICE_CLASS_EMPTY,
+                state_class=STATE_CLASS_MEASUREMENT,
+                unit_of_measurement=UNIT_EMPTY,
+            ).extend({cv.Required(CONF_COMMAND): cv.hex_uint16_t}),
+            cv.Optional(CONF_CUSTOM_5): sensor.sensor_schema(
+                accuracy_decimals=1,
+                device_class=DEVICE_CLASS_EMPTY,
+                state_class=STATE_CLASS_MEASUREMENT,
+                unit_of_measurement=UNIT_EMPTY,
+            ).extend({cv.Required(CONF_COMMAND): cv.hex_uint16_t}),
```

Comment:
Something like this would allow the user to specify a list of any number of custom sensors. It just has to be handled in `to_code` and in the cpp to have `std::vector<sensor::Sensor *> custom_sensors` to handle them.

```suggestion
            cv.Optional(CONF_CUSTOM): cv.ensure_list(
                sensor.sensor_schema(
                    accuracy_decimals=1,
                    device_class=DEVICE_CLASS_EMPTY,
                    state_class=STATE_CLASS_MEASUREMENT,
                    unit_of_measurement=UNIT_EMPTY,
                ).extend({cv.Required(CONF_COMMAND): cv.hex_uint16_t})
            ),
```

Reply:
Thanks for mentioning this. This is a much neater solution.
Applied the suggested changes and did a full test with the actual hardware.
Doc PR is also updated.